domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,239 @@
+package org.folio.rest.persist;
+
+import io.vertx.core.Vertx;
+import io.vertx.pgclient.PgConnection;
+import io.vertx.sqlclient.Pool;
+import io.vertx.core.Future;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.folio.rest.tools.utils.Envs;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class PostgresConnectionManager {
+  public static final int OBSERVER_INTERVAL = 10000;
+  private static final String LOGGER_LABEL = "CONNECTION MANAGER CACHE STATE";
+  private static final Logger LOG = LogManager.getLogger(PostgresConnectionManager.class);
+  private final Collection<CachedPgConnection> connectionCache = Collections.synchronizedCollection(new ArrayList<>());
+  private final int maxPoolSize;
+  private final ConnectionMetrics connectionMetrics = new ConnectionMetrics();
+  private int lowestReleaseDelayReceived;
+
+  public PostgresConnectionManager() {
+    this.maxPoolSize = getMaxSharedPoolSize();
+  }
+
+  public int getCacheSize() {
+    return connectionCache.size();
+  }
+
+  public void clearCache() {
+    connectionMetrics.clear();
+    connectionCache.clear();
+    LOG.debug("Cleared connection manager");
+  }
+
+  public void tryClose(CachedPgConnection connection) {
+    this.tryAddToCache(connection);
+    this.tryRemoveOldestAvailableConnectionAndClose();
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (! PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    LOG.debug("In shared pool mode");
+    return getOrCreateCachedConnection(pool, schemaName, tenantId);
+  }
+
+  public void setObserver(Vertx vertx,  int connectionReleaseDelay) {
+    setObserver(vertx, connectionReleaseDelay, OBSERVER_INTERVAL);
+  }
+
+  public void setObserver(Vertx vertx, int connectionReleaseDelay, int observerInterval) {
+    if (connectionReleaseDelay == 0) {
+      // Zero means there is no timeout and connections should be kept open forever. See the RMB readme.
+      return;
+    }
+
+    if (this.lowestReleaseDelayReceived == 0 || connectionReleaseDelay < lowestReleaseDelayReceived) {
+      // Since there can be multiple clients, we let whichever one has provided the lowest release delay win.
+      this.lowestReleaseDelayReceived = connectionReleaseDelay;
+    }
+
+    LOG.debug("Setting idle connection timeout observer with interval: {}", observerInterval);
+    vertx.setPeriodic(observerInterval, id -> {
+      LOG.debug("Observer firing:: connectionReleaseDelay: {}, observerInterval: {}, observer id {}",
+          connectionReleaseDelay, observerInterval, id);
+      logCache();
```

Comment:
I think we should consider showing the metrics via the MetricsUtil of RMB and having them show up in JMX/Mbeans. This can be done in another iteration.